.github/workflows/publish.yml

```diff
@@ -113,6 +113,34 @@ jobs:
         env:
           NODE_AUTH_TOKEN: ${{ secrets.NPM_PUBLIC_REGISTRY_TOKEN }}
 
+  publish-crates:
+    if: github.event_name == 'push' && contains(github.ref, 'refs/tags/')
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v3
+
+      - name: Setup rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: nightly
+          default: true
+          override: true
+
+      - name: Install cargo make
+        uses: davidB/rust-cargo-make@v1
+
+      - name: Publish package (cosmos_sdk_grpc_client)
+        working-directory: ./rust/cosmos_sdk_grpc_client
+        run: |
+          echo "🚀 Publishing cargo package"
+          cargo make --env CARGO_API_TOKEN="${{ secrets.CARGO_API_TOKEN }}" -- publish
```

Comment:
Publish will not work since you doesn't have a Makefile.toml into the ./rust/cosmos_sdk_grpc_client directory. 
Also you need to generate proto before publish. 

Two solutions : 
- Use matrix 💊 as suggested by @amimart, add Makefile.toml in each directory to configure the publish step for each cargo library.
- Use the main Makefile.toml : 
  - Add two tasks `publish-cosmos` and `publish-okp4`
  - Update `publish` task by adding dependencies on `publish-cosmos` and `publish-okp4` task.
  - Add dependency on `generate-proto` task (as suggested further down in my review) on` publish-cosmos` and `publish-okp4`

Reply:
Mmmh 🤔 I don't know. We could consider use the workspace system of cargo to make the publish from the root folder, If the version is set on the root package, I hope cargo throw an error if version is not the same on each package or maybe no need to set a version on child package if set in the root package. 

Another advantage of using workspace, [all dependencies](https://doc.rust-lang.org/cargo/reference/workspaces.html#key-features) are the same in each package. I think also if you configure the Makefile.toml for publish, the publish will be executed on each child libraries. 

```toml
[package]
name = "proto-build"
version = "1.2.0"
authors = [""]
edition = "2022"
publish = false

[workspace]
members = ["cosmos_sdk_grpc_client", "okp4_grpc_client"]

[dependencies]
prost = "0.10"
prost-build = "0.10"
tonic = "0.7"
tonic-build = "0.7"
regex = "1"
walkdir = "2"

```

Reply:
Yes but by doing this we will include both prost-build and tonic-build dependencies, but they are not needed in the generated and published version

Reply:
@amimart @bdeneux is it ok if I do using the matrix solution ?

Reply:
@warstrolo @bdeneux I think if we can mutualize and avoid duplication we must do it, we should consider putting the `tonic-build` and `prost-build` in the build dependencies instead, this is by the way more relevant in my opinion.

Regarding the version of the root `Cargo.toml`, I hope it can be enough to handle the version of each library, in the opposite we shall handle the version in the 3 files.

Reply:
Build dependencies are only used in the build steps, but since it's our script that do the generation and not the build script of rust (cargo build) idk if that would work.

Btw seems that the root `Cargo.toml` can't define version for subs.

Reply:
@warstrolo Independent of the `workspace` configuration, you can configure a publication of the two cargo in the root `Makefile.toml` by setting two publish tasks : 

```toml
[tasks.publish-cosmos]
args = ["publish", "--token", "${CARGO_API_TOKEN}", "--manifest-path", "cosmos_sdk_grpc_client/Cargo.toml"]
command = "cargo"

[tasks.publish-okp4]
args = ["publish", "--token", "${CARGO_API_TOKEN}", "--manifest-path", "okp4_grpc_client/Cargo.toml"]
command = "cargo"

[tasks.publish]
dependencies = ["publish-cosmos", "publish-okp4"]
```

Reply:
@warstrolo you can define version for dependencies inside each members but Cargo will ensure that all dependencies in the same workspace are on the same version. https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch14-03-cargo-workspaces.html#depending-on-an-external-crate-in-a-workspace